### PR TITLE
kernel/processbuffer: add Miri tests for aliasing & provenance

### PR DESCRIPTION
**Pull request by lschuermann**
### Pull Request Overview

This adds a bunch of Miri tests (which were initially created by Gemini 3-Pro, and then heavily modified) to exercise various edge cases of the `ProcessSlice`-related infrastructure in `processbuffer.rs`.

These types and their associated helper methods perform a bunch of operations that use unsafe, and could be problematic given Rust's restrictions on reference validity, aliasing, and provenance. As such, these tests are designed to exercise various edge cases, such as re-tagging of pointers, creating references over null pointers, concurrent accesses through aliased references, and out-of-bounds accesses.

In particular, this includes a test case that uncovers unsound aliasing in one of the helper methods of this file (fix in https://github.com/tock/tock/pull/4700), and one test case exercising the previously known issue of `impl From<&[u8]> for ReadableProcessSlice` being unsound under stacked borrows.

### Testing Strategy

Tests itself, using

```
$ MIRIFLAGS="-Zmiri-strict-provenance -Zmiri-symbolic-alignment-check -Zmiri-retag-fields -Zmiri-tree-borrows" \
    RUSTFLAGS="-Z randomize-layout" \
    cargo miri test
```

### TODO or Help Wanted

TODO: are we running these checks in CI?

I think this and other recent CI PRs, in addition to the Treadmill integration show that our current strategy of running checks in CI is becoming unwiedly and reaching its limits. We are having a hard time keeping track of which checks are actually run, which of them cause actual CI errors, and we are running a bunch of checks even for PRs that couldn't possibly affect them.

Maybe this PR is a good motivator to try and revise the way we're doing CI for Tock more broadly.

### Documentation Updated

- [X] ~Updated the relevant files in `/docs`,~ or no updates are required.

### Formatting

- [X] Ran `make prepush`.

### CODE REVIEW

File: kernel/src/processbuffer.rs [hunk 1]
```diff
@@ -1157,3 +1157,332 @@ impl<I: ProcessSliceIndex<Self>> Index<I> for WriteableProcessSlice {
         index.index(self)
     }
 }
+
+#[cfg(test)]
+mod miri_tests {
```
Comment (bradjc) [inline kernel/src/processbuffer.rs:1162]: Does this name `miri_tests` have any meaning to rust/cargo?

I think it makes sense to try to group tests based on how they are evaluated, but I'm wondering what the line is between requirement from the toolchain(s) vs. tock convention.

Comment (lschuermann) [reply kernel/src/processbuffer.rs:1162]: It seemed appropriate to call this `miri_tests`, as they are a little toothless when not run under Miri (albeit still work as regular Cargo tests). I don't know what the convention is, I'm happy to just call it `tests` and add a comment that these are intended to be run under Miri (with a certain set of flags)?

**Comment by lschuermann** [comment]
Marked as ready-for-review, I doubt that these actually run in CI under Miri, but that's probably best solved in a separate follow-up PR.

File: kernel/src/processbuffer.rs (hunk 1 above)
Comment (bradjc) [reply kernel/src/processbuffer.rs:1162]: Yeah let's keep it as `miri_tests`. We can eventually try to standardize how we include normal cargo tests, miri tests, and flux checks at the bottom of files.

It seems generally helpful for future readers to understand that the miri-focused tests aren't really trying to test the implementation functionality.
